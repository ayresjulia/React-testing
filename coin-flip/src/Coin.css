.Coin-img {
	width: 200px;
	height: 200px;
}

.Coin-btn {
	margin-top: 10px;
}
