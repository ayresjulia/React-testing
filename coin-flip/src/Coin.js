import React, { useState } from "react";
import _ from "lodash";
import "./Coin.css";

const Coin = (props) => {
	const [ coin, setCoin ] = useState(_.shuffle(props.coin_url)[0]);
	const [ count, setCount ] = useState(0);
	const [ head, setHead ] = useState(0);
	const [ tail, setTail ] = useState(0);

	const flip = () => {
		let shuffle = _.shuffle(props.coin_url)[0];
		setCoin(shuffle);
		shuffle === props.coin_url[0] ? setHead(head + 1) : setTail(tail + 1);
		setCount(count + 1);
	};
	// coin is showing only when count > 0
	const isShowing = count !== 0;

	return (
		<div className="Coin">
			<h1>Let's flip a coin!</h1>
			{isShowing && <img data-testid="coin-img" className="Coin-img" src={coin} alt="" />}
			<div>
				<button data-testid="coin-btn" className="Coin-btn" onClick={flip}>
					FLIP ME
				</button>
			</div>
			<p data-testid="coin-count">
				Out of {count} flips, there have been {head} heads and {tail} tails.
			</p>
		</div>
	);
};

Coin.defaultProps = {
	coin_url: [
		"https://i.ebayimg.com/images/g/xtcAAOSwLwBaZigS/s-l400.jpg",
		"https://i.ebayimg.com/images/g/wGEAAOSwYu1crzzn/s-l400.jpg"
	]
};

export default Coin;
