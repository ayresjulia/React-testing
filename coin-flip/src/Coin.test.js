import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Coin from "./Coin";

// use for random flip of a coin and score count
const heads = "https://i.ebayimg.com/images/g/xtcAAOSwLwBaZigS/s-l400.jpg";

// smoke test
it("renders without crashing", () => {
	render(<Coin />);
});

// snapshot test
it("matches snapshot", () => {
	// asFragment gives us an object
	const { asFragment } = render(<Coin />);
	expect(asFragment()).toMatchSnapshot();
});

it("shows the coin when you click on the the button", () => {
	const { queryByTestId } = render(<Coin />);
	const coinImg = queryByTestId("coin-img");

	// expect the coin image to not be seen with zero clicks
	expect(coinImg).not.toBeInTheDocument();
});

it("increments count of either  head or tail on flip of a coin", () => {
	const { getByTestId, queryByTestId } = render(<Coin />);

	// get button and click
	const coinBtn = queryByTestId("coin-btn");
	fireEvent.click(coinBtn);

	// get content of scores string
	const checkContent = getByTestId("coin-count");

	// get contents of coin-img src and compare to expected scores
	let imgSrc = queryByTestId("coin-img").src;
	imgSrc === heads
		? expect(checkContent).toHaveTextContent("Out of 1 flips, there have been 1 heads and 0 tails.")
		: expect(checkContent).toHaveTextContent("Out of 1 flips, there have been 0 heads and 1 tails.");

	// coin-img in the document
	expect(queryByTestId("coin-img")).toBeInTheDocument();
});
