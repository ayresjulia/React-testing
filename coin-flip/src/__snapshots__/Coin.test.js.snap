// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Coin"
  >
    <h1>
      Let's flip a coin!
    </h1>
    <div>
      <button
        class="Coin-btn"
        data-testid="coin-btn"
      >
        FLIP ME
      </button>
    </div>
    <p
      data-testid="coin-count"
    >
      Out of 0 flips, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
